core/src/main/kotlin/rys/ajaxpetproject/service/impl/UserServiceReactiveImpl.kt

```diff
@@ -0,0 +1,98 @@
+package rys.ajaxpetproject.service.impl
+
+import org.springframework.security.crypto.password.PasswordEncoder
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+import rys.ajaxpetproject.exceptions.UserAlreadyExistsException
+import rys.ajaxpetproject.exceptions.UserNotFoundException
+import rys.ajaxpetproject.model.MongoUser
+import rys.ajaxpetproject.repository.UserRepository
+import rys.ajaxpetproject.service.UserService
+
+@Service
+@Suppress("MagicNumber")
+class UserServiceReactiveImpl(
+    private val encoder: PasswordEncoder,
+    private val userRepository: UserRepository
+) : UserService {
+    override fun createUser(mongoUser: MongoUser): Mono<MongoUser> {
+        return userRepository.findByName(mongoUser.userName)
+            .flatMap {
+                Mono.error<MongoUser>(
+                    UserAlreadyExistsException("A user with the username ${mongoUser.userName} already exists!")
+                )
+            }
```

Comment:
As I see handle here doing : 
1) We can directly reach actual user
2)We sending error to downstream
3)ideologically flatMap can't be used because we don't transform recieved item 

Am I right?

Reply:
Your and mine approaches give the same result. But my approach is more lightweight

Reply:
Thanks!
Really good catch!